### ISSUE TITLE
Spurious abstraction violation with eta contracted functions

### ISSUE THREAD

**Issue by catalin-hritcu** [FStarLang/FStar#1427]
Some of the code I used to have on summer school slides started breaking:
http://prosecco.gforge.inria.fr/personal/hritcu/teaching/aarhus2017/slides/out/pure-fun/pure-fun.html#/sec-using-refinement-types-for-stacks

It works with `--use_two_phase_tc false` but fails with `--use_two_phase_tc true`:
```
fstar.exe --use_two_phase_tc true  RefinedStack.fst
RefinedStack.fst(11,62-11,63): (Error 189) Interface of RefinedStack violates its abstraction (add a 'private' qualifier to 'push'?): Expected expression of type "Prims.list Prims.int"; got expression "s" of type "RefinedStack.stack" (see also RefinedStack.fst(11,43-11,44))
```

Writing `let is_empty s = Nil? s` instead of `let is_empty = Nil?` makes it work.

**Comment by aseemr** [comment]
This is not a two phase tc problem. I can reproduce it even in the single phase:

```
module Test

abstract type stack = list int
  
abstract val is_empty : stack -> Tot bool
let is_empty = Nil? #int

abstract val empty : s:stack{is_empty s}
let empty = []
```

Note the implicit argument `#int` to `Nil?` in `is_empty`.

It comes up in two-phases because the first phase adds this argument and then typechecks again.

My guess is: somewhere in checking the application of `Nil?`, we are unfolding the expected type `stack -> Tot bool` to `list int -> Tot bool`, and that gets annotated on the let-binding.

**Comment by catalin-hritcu** [comment]
Did one more test and it seems there is nothing special about the projector here. I can define my own and the problem still happens:
```
module Bug1427

abstract type stack = list int

abstract val is_nil : #a:Type -> list a -> Tot bool
let is_nil #a xs =
  match xs with
  | Nil -> true
  | Cons _ _ -> false

abstract val is_empty : stack -> Tot bool
let is_empty = is_nil #int

abstract val empty : s:stack{is_empty s}
let empty = []
```

**Comment by aseemr** [comment]
Here is the problem:

When typecheking `is_empty`, we first typecheck the body of the function `Nil? #int` as `list int -> Tot bool`. Then, we weaken the type as per the annotated type `stack -> Tot bool`. In this case, the weakening is trivial (a simple unfolding of `stack` does the trick), and so, the next step is to set the type of the `let` binding as `stack -> Tot bool`.

This is where the current code has a bug:

https://github.com/FStarLang/FStar/blob/master/src/typechecker/FStar.TypeChecker.Util.fs#L1132

We set the `res_typ` in the lazy comp `lc`, not doing anything for the embedded `comp` thunk inside it (which still has the return type `list int -> Tot bool`). Down the line, the code reduces this `comp` thunk and annotates the `let` binding with the `comp` result type. And so, the `lc.res_typ` is effectively ignored. For the rest of the module, `is_empty` has type `list int -> bool`, and we try to apply it to a `stack`, which is `abstract`, this gives an error when checking for abstraction violations.

It works when the implicit argument is not given, because in that case, there is a `uvar` in the `wp` of the body, which means that the weakening is not trivial and we don't go into this branch of `weaken_result_typ`.

(More to follow.)

**Comment by aseemr** [comment]
The fix for this problem is to set the result type in the `comp` thunk inside `lcomp` as well. I tried that fix, but it breaks some other examples, which were incorrectly going through earlier.

One such instance is examples of the following form:

```
let f () = e <: Lemma phi
```
In this case, we first typecheck `e` to let's say `_:unit{psi}`. At this point, the code checks *and sets* the expected *type* to `unit`. After that it checks for the expected effect, trying to prove `Tot unit <: Lemma phi`, which it is unable to prove (https://github.com/FStarLang/FStar/blob/master/src/typechecker/FStar.TypeChecker.TcTerm.fs#L378),

This used to work earlier since the code (incorrectly) ignored the setting of the result type in the `comp` thunk. So effectively we were checking `Tot _:unit{psi} <: Lemma phi`, which could be provable.

The fix for this problem is to not check the expected type separately, rather just check the expected effect. With that fix, I noticed some other regressions, both in smt and in unification.

Need to debug more.
